src/Features/Core/Portable/Diagnostics/DiagnosticService.cs

```diff
@@ -114,12 +114,28 @@ private bool UpdateDataMap(IDiagnosticUpdateSource source, DiagnosticsUpdatedArg
                     return false;
                 }
 
-                var diagnosticDataMap = _map.GetOrAdd(source, _ => new Dictionary<object, Data>());
+                // 2 different workspaces (ex, PreviewWorkspaces) can return same Args.Id, we need to
+                // distinguish them. so we separate diagnostics per workspace map.
+                var workspaceMap = _map.GetOrAdd(source, _ => new Dictionary<Workspace, Dictionary<object, Data>>());
+
+                if (args.Diagnostics.Length == 0 && !workspaceMap.ContainsKey(args.Workspace))
+                {
+                    // no new diagnostic, and we don't have update source for it.
+                    return false;
+                }
+
+                var diagnosticDataMap = workspaceMap.GetOrAdd(args.Workspace, _ => new Dictionary<object, Data>());
 
                 diagnosticDataMap.Remove(args.Id);
                 if (diagnosticDataMap.Count == 0 && args.Diagnostics.Length == 0)
                 {
-                    _map.Remove(source);
+                    workspaceMap.Remove(args.Workspace);
```

Comment:
it's slightly odd that we would call "GetOrAdd" (thus potentially creating a key/value pair) only to then remove it right here.